README.md
# CropperX

A cropper package made fully in Flutter.

  Circle                   |  Rectangle                |  Grid
:-------------------------:|:-------------------------:|:-------------------------:
![Screenshot circle](doc/screenshot_circle.png)|![Screenshot rectangle](doc/screenshot_rectangle.png)|![Screenshot grid](doc/screenshot_grid.png)

### Types of overlays

The overlay can be defined by adding an overlay type through the `overlayType` parameter. The following types are currently supported:

* **OverlayType.none**: No overlay at all.
* **OverlayType.circle**: An overlay with a circular gap.
* **OverlayType.rectangle**: An overlay with a rectangular gap, takes the given aspect ratio.
* **OverlayType.grid**: An overlay of grid lines: 2 horizontal and 2 vertical, evenly spaced.
* **OverlayType.gridHorizontal**: An overlay of 2 horizontal grid lines, evenly spaced.
* **OverlayType.gridVertical**: An overlay of 2 vertical grid lines, evenly spaced.

### Rotating the image

The image rotation can be defined through the `rotationTurns` parameter. The default value is `0`, corresponding to 0 degrees of rotation. Making the value larger by 1 will rotate the image 90 degrees clockwise, making it smaller will rotate the image 90 degrees counterclockwise. `0` to `4` is not a limit, so going above or below these values will keep rotating the image.

### Example

```dart
// Import package
import 'package:cropperx/cropper.dart';

// Define a key
final _cropperKey = GlobalKey(debugLabel: 'cropperKey');

// Add the Cropper widget to your tree, using the above key and adding the image to crop
Cropper(
    cropperKey: _cropperKey, // Use your key here
    image: Image.network(
        'https://i.pinimg.com/originals/6b/4d/18/6b4d18c0b756ab20c3591490dfc10090.jpg',
    ),
)

// Get the cropped image as bytes
final imageBytes = await Cropper.crop(
    cropperKey: _cropperKey, // Reference it through the key
)
```

CHANGELOG.md
## 1.1.1 - 2022/07/14
* Fix weird gesture handling when image was rotated
* Fix image placement after rotation

## 1.1.0 - 2022/05/21
* Add Flutter 3 support
* Fix overlay not filling viewport
* Center image on initial image load

## 1.0.0 - 2021/12/07
* Fix Rectangle overlay not showing

## 0.0.3 - 2021/10/08
* Make pixelRatio configurable when calling `crop`

## 0.0.2 - 2021/10/05
* Add Flutter Web support
* Fix incorrect circle frame when parent's width exceeds height

## 0.0.1 - 2021/10/04
* Initial release

test/flutter_cropperx_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_cropperx/cropper.dart';

void main() {
  // TODO
}


lib/cropper.dart
import 'dart:io';
import 'dart:typed_data';
import 'dart:ui';

import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';

class Cropper extends StatefulWidget {
  /// The cropper's key to reference when calling the crop function.
  final GlobalKey? cropperKey;

  /// The background color of the cropper widget, visible when the image won't
  /// fill the entire widget. Defaults to a light grey color: Color(0xFFCECECE).
  final Color backgroundColor;

  /// The color of the cropper's overlay. Defaults to semi-transparent black
  /// Colors.black54
  final Color overlayColor;

  /// The type of semi-transparent overlay. Can either be an
  /// [OverlayType.circle] or [OverlayType.none] to hide the overlay. Defaults
  /// to none so no overlay is shown by default.
  final OverlayType overlayType;

  /// The maximum scale the user is able to zoom. Defaults to 2.5
  final double zoomScale;

  /// The aspect ratio to crop the image to. Defaults to a square (an aspect ratio of 1.0)
  final double aspectRatio;

  /// The number of clockwise quarter turns the image should be rotated. Defaults to 0.
  final int rotationTurns;

  /// The thickness of the grid lines. Defaults to 2.0.
  final double gridLineThickness;

  /// The image to crop.
  final Image image;

  const Cropper({
    Key? key,
    this.backgroundColor = const Color(0xFFCECECE),
    this.overlayColor = Colors.black38,
    this.overlayType = OverlayType.none,
    this.zoomScale = 2.5,
    this.gridLineThickness = 2.0,
    this.aspectRatio = 1,
    this.rotationTurns = 0,
    required this.cropperKey,
    required this.image,
  }) : super(key: key);

  @override
  State<Cropper> createState() => _CropperState();

  /// Crops the image as displayed in the cropper widget. The cropper widget should be referenced
  /// using its key.
  static Future<File> crop({
    required GlobalKey cropperKey,
    required String path,
    required String fileName,
  }) async {
    // Get cropped image
    final boundary = cropperKey.currentContext!.findRenderObject()! as RenderRepaintBoundary;
    final image = await boundary.toImage(pixelRatio: 3);

    // Convert image to bytes in PNG format
    final ByteData? byteData = await image.toByteData(format: ImageByteFormat.png);
    final Uint8List pngBytes = byteData!.buffer.asUint8List();

    // Create and return file
    return File('$path/$fileName.png').writeAsBytes(pngBytes);
  }
}

class _CropperState extends State<Cropper> {
  late final TransformationController _transformationController;

  /// Boolean to indicate if the image has been updated after a state change. Used so we don't do
  /// any unnecessary refreshes.
  late bool _hasImageUpdated;

  /// Boolean to indicate whether we need to set the initial scale of an image.
  bool _shouldSetInitialScale = false;

  /// The image configuration used to add the image stream listener to the image.
  final _imageConfiguration = const ImageConfiguration();

  /// Image stream listener which is used to indicate whether the image has finished loading. This
  /// is required to do the initial scaling of the [InteractiveViewer], where we'd like to fill the
  /// viewport by scaling the image down as much as possible.
  late final _imageStreamListener = ImageStreamListener(
    (_, __) {
      setState(() {
        _shouldSetInitialScale = true;
      });
    },
  );

  @override
  void initState() {
    super.initState();
    _hasImageUpdated = true;
    _transformationController = TransformationController();
  }

  @override
  void didUpdateWidget(covariant Cropper oldWidget) {
    super.didUpdateWidget(oldWidget);
    _hasImageUpdated = oldWidget.image.image != widget.image.image;
  }

  @override
  Widget build(BuildContext context) {
    return ClipRect(
      child: ColoredBox(
        color: widget.backgroundColor,
        child: Stack(
          alignment: Alignment.center,
          children: [
            RepaintBoundary(
              key: widget.cropperKey,
              child: RotatedBox(
                quarterTurns: widget.rotationTurns,
                child: AspectRatio(
                  aspectRatio: widget.aspectRatio,
                  child: LayoutBuilder(
                    builder: (_, constraint) {
                      return InteractiveViewer(
                        clipBehavior: Clip.none,
                        transformationController: _transformationController,
                        constrained: false,
                        child: Builder(
                          builder: (context) {
                            final imageStream = widget.image.image.resolve(_imageConfiguration);
                            if (_hasImageUpdated && _shouldSetInitialScale) {
                              imageStream.removeListener(_imageStreamListener);
                              WidgetsBinding.instance?.addPostFrameCallback((_) {
                                final renderBox = context.findRenderObject() as RenderBox?;
                                final childSize = renderBox?.size ?? Size.zero;
                                if (childSize != Size.zero) {
                                  final parentSize = constraint.biggest;
                                  _transformationController.value =
                                      Matrix4.identity() * _getCoverRatio(parentSize, childSize);
                                }
                              });
                            }

                            if (_hasImageUpdated && !_shouldSetInitialScale) {
                              imageStream.addListener(_imageStreamListener);
                            }

                            return widget.image;
                          },
                        ),
                        minScale: 0.1,
                        maxScale: widget.zoomScale,
                      );
                    },
                  ),
                ),
              ),
            ),
            if (widget.overlayType == OverlayType.circle)
              ClipPath(
                clipper: _OverlayCircleFrame(),
                child: Container(
                  color: widget.overlayColor,
                ),
              ),
            if (widget.overlayType == OverlayType.rectangle)
              ClipPath(
                clipper: _OverlayRectangleFrame(
                  aspectRatio: widget.aspectRatio,
                ),
                child: Container(
                  color: widget.overlayColor,
                ),
              ),
            if (widget.overlayType == OverlayType.grid ||
                widget.overlayType == OverlayType.gridHorizontal)
              Column(
                mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                children: [
                  Divider(
                    color: widget.overlayColor,
                    thickness: widget.gridLineThickness,
                  ),
                  Divider(
                    color: widget.overlayColor,
                    thickness: widget.gridLineThickness,
                  ),
                ],
              ),
            if (widget.overlayType == OverlayType.grid ||
                widget.overlayType == OverlayType.gridVertical)
              Row(
                mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                children: [
                  VerticalDivider(
                    color: widget.overlayColor,
                    thickness: widget.gridLineThickness,
                  ),
                  VerticalDivider(
                    color: widget.overlayColor,
                    thickness: widget.gridLineThickness,
                  ),
                ],
              ),
          ],
        ),
      ),
    );
  }

  double _getCoverRatio(Size outside, Size inside) {
    return outside.width / outside.height > inside.width / inside.height
        ? outside.width / inside.width
        : outside.height / inside.height;
  }

  @override
  void dispose() {
    _transformationController.dispose();
    super.dispose();
  }
}

enum OverlayType { circle, rectangle, grid, gridHorizontal, gridVertical, none }

class _OverlayCircleFrame extends CustomClipper<Path> {
  @override
  Path getClip(Size size) {
    return Path.combine(
      PathOperation.difference,
      Path()..addRect(Rect.fromLTWH(0, 0, size.width, size.height)),
      Path()
        ..addOval(
          Rect.fromCircle(
            center: Offset(size.width / 2, size.height / 2),
            radius: size.width / 2,
          ),
        )
        ..close(),
    );
  }

  @override
  bool shouldReclip(CustomClipper<Path> oldClipper) => true;
}

class _OverlayRectangleFrame extends CustomClipper<Path> {
  final double aspectRatio;

  _OverlayRectangleFrame({
    required this.aspectRatio,
  });

  @override
  Path getClip(Size size) {
    double _height = aspectRatio >= 1 ? size.width / aspectRatio : size.height;
    double _width = aspectRatio <= 1 ? size.height * aspectRatio : size.width;

    return Path.combine(
      PathOperation.difference,
      Path()..addRect(Rect.fromLTWH(0, 0, size.width, size.height)),
      Path()
        ..addRect(
          Rect.fromCenter(
            center: Offset(size.width / 2, size.height / 2),
            height: _height,
            width: _width,
          ),
        )
        ..close(),
    );
  }

  @override
  bool shouldReclip(CustomClipper<Path> oldClipper) =>
      aspectRatio != (oldClipper as _OverlayRectangleFrame).aspectRatio;
}


lib/cropperx.dart
import 'dart:math';
import 'dart:typed_data';
import 'dart:ui';

import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';

class Cropper extends StatefulWidget {
  /// The cropper's key to reference when calling the crop function.
  final GlobalKey? cropperKey;

  /// The background color of the cropper widget, visible when the image won't
  /// fill the entire widget. Defaults to a light grey color: Color(0xFFCECECE).
  final Color backgroundColor;

  /// The color of the cropper's overlay. Defaults to semi-transparent black
  /// Colors.black54
  final Color overlayColor;

  /// The type of semi-transparent overlay. Can either be an
  /// [OverlayType.circle] or [OverlayType.none] to hide the overlay. Defaults
  /// to none so no overlay is shown by default.
  final OverlayType overlayType;

  /// The maximum scale the user is able to zoom. Defaults to 2.5
  final double zoomScale;

  /// The aspect ratio to crop the image to. Defaults to a square (an aspect ratio of 1.0)
  final double aspectRatio;

  /// The number of clockwise quarter turns the image should be rotated. Defaults to 0.
  final int rotationTurns;

  /// The thickness of the grid lines. Defaults to 2.0.
  final double gridLineThickness;

  /// The image to crop.
  final Image image;

  /// The called when scale start.
  final GestureScaleStartCallback? onScaleStart;

  /// The called when scale update.
  final GestureScaleUpdateCallback? onScaleUpdate;

  /// The called when scale end.
  final GestureScaleEndCallback? onScaleEnd;

  const Cropper({
    Key? key,
    this.backgroundColor = const Color(0xFFCECECE),
    this.overlayColor = Colors.black38,
    this.overlayType = OverlayType.none,
    this.zoomScale = 2.5,
    this.gridLineThickness = 2.0,
    this.aspectRatio = 1,
    this.rotationTurns = 0,
    this.onScaleStart,
    this.onScaleUpdate,
    this.onScaleEnd,
    required this.cropperKey,
    required this.image,
  }) : super(key: key);

  @override
  State<Cropper> createState() => _CropperState();

  /// Crops the image as displayed in the cropper widget, converts it to PNG format and returns it
  /// as [Uint8List]. The cropper widget should be referenced using its key.
  static Future<Uint8List?> crop({
    required GlobalKey cropperKey,
    double pixelRatio = 3,
  }) async {
    // Get cropped image
    final renderObject = cropperKey.currentContext!.findRenderObject();
    final boundary = renderObject as RenderRepaintBoundary;
    final image = await boundary.toImage(pixelRatio: pixelRatio);

    // Convert image to bytes in PNG format and return
    final byteData = await image.toByteData(
      format: ImageByteFormat.png,
    );
    final pngBytes = byteData?.buffer.asUint8List();

    return pngBytes;
  }
}

class _CropperState extends State<Cropper> {
  late final TransformationController _transformationController;

  /// Boolean to indicate if the image has been updated after a state change. Used so we don't do
  /// any unnecessary refreshes.
  late bool _hasImageUpdated;

  /// Boolean to indicate whether we need to set the initial scale of an image.
  bool _shouldSetInitialScale = false;
  bool _shouldUpdateScale = false;

  /// The image configuration used to add the image stream listener to the image.
  final _imageConfiguration = const ImageConfiguration();

  /// Image stream listener which is used to indicate whether the image has finished loading. This
  /// is required to do the initial scaling of the [InteractiveViewer], where we'd like to fill the
  /// viewport by scaling the image down as much as possible.
  late final _imageStreamListener = ImageStreamListener(
    (_, __) {
      WidgetsBinding.instance.addPostFrameCallback((_) {
        setState(() {
          _shouldSetInitialScale = true;
        });
      });
    },
  );

  @override
  void initState() {
    super.initState();
    _hasImageUpdated = true;
    _transformationController = TransformationController();
  }

  @override
  void didUpdateWidget(covariant Cropper oldWidget) {
    super.didUpdateWidget(oldWidget);
    _hasImageUpdated = oldWidget.image.image != widget.image.image;
    _shouldUpdateScale = oldWidget.rotationTurns != widget.rotationTurns;
  }

  @override
  Widget build(BuildContext context) {
    return ClipRect(
      child: ColoredBox(
        color: widget.backgroundColor,
        child: Stack(
          alignment: Alignment.center,
          children: [
            RepaintBoundary(
              key: widget.cropperKey,
              child: AspectRatio(
                aspectRatio: widget.aspectRatio,
                child: LayoutBuilder(
                  builder: (_, constraint) {
                    return InteractiveViewer(
                      clipBehavior: Clip.none,
                      transformationController: _transformationController,
                      constrained: false,
                      child: Builder(
                        builder: (context) {
                          final imageStream = widget.image.image.resolve(
                            _imageConfiguration,
                          );
                          if (_hasImageUpdated && _shouldSetInitialScale) {
                            imageStream.removeListener(_imageStreamListener);
                            _setInitialScale(context, constraint.biggest);
                          }

                          if (_hasImageUpdated && !_shouldSetInitialScale) {
                            imageStream.addListener(_imageStreamListener);
                          }

                          if (_shouldUpdateScale) {
                            _setInitialScale(context, constraint.biggest);
                            _shouldUpdateScale = false;
                          }

                          return RotatedBox(
                            quarterTurns: widget.rotationTurns,
                            child: widget.image,
                          );
                        },
                      ),
                      minScale: 0.1,
                      maxScale: widget.zoomScale,
                      onInteractionStart: widget.onScaleStart,
                      onInteractionUpdate: widget.onScaleUpdate,
                      onInteractionEnd: widget.onScaleEnd,
                    );
                  },
                ),
              ),
            ),
            if (widget.overlayType == OverlayType.circle ||
                widget.overlayType == OverlayType.rectangle)
              Positioned.fill(
                child: ClipPath(
                  clipper: _OverlayFrame(
                    aspectRatio: widget.aspectRatio,
                    isCircle: widget.overlayType == OverlayType.circle,
                  ),
                  child: Container(
                    color: widget.overlayColor,
                  ),
                ),
              ),
            if (widget.overlayType == OverlayType.grid ||
                widget.overlayType == OverlayType.gridHorizontal)
              Positioned.fill(
                child: Column(
                  mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                  children: [
                    Divider(
                      color: widget.overlayColor,
                      thickness: widget.gridLineThickness,
                    ),
                    Divider(
                      color: widget.overlayColor,
                      thickness: widget.gridLineThickness,
                    ),
                  ],
                ),
              ),
            if (widget.overlayType == OverlayType.grid ||
                widget.overlayType == OverlayType.gridVertical)
              Positioned.fill(
                child: Row(
                  mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                  children: [
                    VerticalDivider(
                      color: widget.overlayColor,
                      thickness: widget.gridLineThickness,
                    ),
                    VerticalDivider(
                      color: widget.overlayColor,
                      thickness: widget.gridLineThickness,
                    ),
                  ],
                ),
              ),
          ],
        ),
      ),
    );
  }

  double _getCoverRatio(Size outside, Size inside) {
    return outside.width / outside.height > inside.width / inside.height
        ? outside.width / inside.width
        : outside.height / inside.height;
  }

  double _getTranslationX(Size outside, Size inside, coverRatio) {
    return (outside.width / coverRatio - inside.width) / 2;
  }

  double _getTranslationY(Size outside, Size inside, coverRatio) {
    return (outside.height / coverRatio - inside.height) / 2;
  }

  void _setInitialScale(BuildContext context, Size parentSize) {
    WidgetsBinding.instance.addPostFrameCallback((_) {
      final renderBox = context.findRenderObject() as RenderBox?;
      final childSize = renderBox?.size ?? Size.zero;
      if (childSize != Size.zero) {
        final coverRatio = _getCoverRatio(parentSize, childSize);
        final value = Matrix4.identity() * coverRatio;

        // Center the image inside the InteractiveViewer
        value.translate(
          _getTranslationX(parentSize, childSize, coverRatio),
          _getTranslationY(parentSize, childSize, coverRatio),
          0.0,
        );

        _transformationController.value = value;
      }

      _shouldSetInitialScale = false;
    });
  }

  @override
  void dispose() {
    _transformationController.dispose();
    super.dispose();
  }
}

enum OverlayType { circle, rectangle, grid, gridHorizontal, gridVertical, none }

class _OverlayFrame extends CustomClipper<Path> {
  final double aspectRatio;
  final bool isCircle;

  _OverlayFrame({
    required this.aspectRatio,
    this.isCircle = false,
  });

  @override
  Path getClip(Size size) {
    double _height = aspectRatio >= 1 ? size.width / aspectRatio : size.height;
    double _width = aspectRatio <= 1 ? size.height * aspectRatio : size.width;

    final opening = Path();
    if (isCircle) {
      opening.addOval(
        Rect.fromCircle(
          center: Offset(size.width / 2, size.height / 2),
          radius: min(_height, _width) / 2,
        ),
      );
    } else {
      opening.addRect(
        Rect.fromCenter(
          center: Offset(size.width / 2, size.height / 2),
          height: _height,
          width: _width,
        ),
      );
    }

    return Path.combine(
      PathOperation.difference,
      Path()..addRect(Rect.fromLTWH(0, 0, size.width, size.height)),
      opening..close(),
    );
  }

  @override
  bool shouldReclip(CustomClipper<Path> oldClipper) =>
      aspectRatio != (oldClipper as _OverlayFrame).aspectRatio;
}


example/lib/main.dart
import 'dart:typed_data';

import 'package:cropperx/cropperx.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:image_picker/image_picker.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'CropperX Example',
      theme: ThemeData(
        primaryColor: Colors.blueAccent,
      ),
      home: const CropperScreen(),
    );
  }
}

class CropperScreen extends StatefulWidget {
  const CropperScreen({Key? key}) : super(key: key);

  @override
  State<CropperScreen> createState() => _CropperScreenState();
}

class _CropperScreenState extends State<CropperScreen> {
  final ImagePicker _picker = ImagePicker();
  final GlobalKey _cropperKey = GlobalKey(debugLabel: 'cropperKey');
  Uint8List? _imageToCrop;
  Uint8List? _croppedImage;
  OverlayType _overlayType = OverlayType.circle;
  int _rotationTurns = 0;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SafeArea(
        child: SingleChildScrollView(
          child: Column(
            children: [
              SizedBox(
                height: 500,
                child: _imageToCrop != null
                    ? Cropper(
                        cropperKey: _cropperKey,
                        overlayType: _overlayType,
                        rotationTurns: _rotationTurns,
                        image: Image.memory(_imageToCrop!),
                        onScaleStart: (details) {
                          // todo: define started action.
                        },
                        onScaleUpdate: (details) {
                          // todo: define updated action.
                        },
                        onScaleEnd: (details) {
                          // todo: define ended action.
                        },
                      )
                    : const ColoredBox(color: Colors.grey),
              ),
              const SizedBox(height: 16),
              Wrap(
                spacing: 16,
                children: [
                  ElevatedButton(
                    child: const Text('Pick image'),
                    onPressed: () async {
                      final image = await _picker.pickImage(
                        source: ImageSource.gallery,
                      );

                      if (image != null) {
                        final imageBytes = await image.readAsBytes();
                        setState(() {
                          _imageToCrop = imageBytes;
                        });
                      }
                    },
                  ),
                  ElevatedButton(
                    child: const Text('Switch overlay'),
                    onPressed: () {
                      setState(() {
                        _overlayType = _overlayType == OverlayType.circle
                            ? OverlayType.grid
                            : _overlayType == OverlayType.grid
                                ? OverlayType.rectangle
                                : OverlayType.circle;
                      });
                    },
                  ),
                  ElevatedButton(
                    child: const Text('Crop image'),
                    onPressed: () async {
                      final imageBytes = await Cropper.crop(
                        cropperKey: _cropperKey,
                      );

                      if (imageBytes != null) {
                        setState(() {
                          _croppedImage = imageBytes;
                        });
                      }
                    },
                  ),
                  ElevatedButton(
                    child: const Text('Reset to initial image'),
                    onPressed: () async {
                      final imageBytes = Cropper.reset(
                        cropperKey: _cropperKey,
                      );
                    },
                  ),
                  IconButton(
                    onPressed: () {
                      setState(() => _rotationTurns--);
                    },
                    icon: const Icon(Icons.rotate_left),
                  ),
                  IconButton(
                    onPressed: () {
                      setState(() => _rotationTurns++);
                    },
                    icon: const Icon(Icons.rotate_right),
                  ),
                ],
              ),
              const SizedBox(height: 16),
              if (_croppedImage != null)
                Padding(
                  padding: const EdgeInsets.all(36.0),
                  child: Image.memory(_croppedImage!),
                ),
            ],
          ),
        ),
      ),
    );
  }
}


example/lib/cropper_screen.dart
import 'dart:io';

import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:cropperx/cropperx.dart';
import 'package:path_provider/path_provider.dart';

class CropperScreen extends StatefulWidget {
  const CropperScreen({Key? key}) : super(key: key);

  @override
  State<CropperScreen> createState() => _CropperScreenState();
}

class _CropperScreenState extends State<CropperScreen> {
  final GlobalKey _cropperKey = GlobalKey(debugLabel: 'cropperKey');
  File? _croppedFile;
  OverlayType _overlayType = OverlayType.circle;
  int _rotationTurns = 0;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SafeArea(
        child: SingleChildScrollView(
          child: Column(
            children: [
              SizedBox(
                height: 500,
                child: Cropper(
                  cropperKey: _cropperKey,
                  overlayType: _overlayType,
                  rotationTurns: _rotationTurns,
                  image: Image.network(
                    'https://i.pinimg.com/originals/6b/4d/18/6b4d18c0b756ab20c3591490dfc10090.jpg',
                  ),
                ),
              ),
              const SizedBox(height: 16),
              Wrap(
                spacing: 16,
                children: [
                  ElevatedButton(
                    child: const Text('Switch overlay'),
                    onPressed: () {
                      setState(() {
                        _overlayType = _overlayType == OverlayType.circle
                            ? OverlayType.grid
                            : _overlayType == OverlayType.grid
                                ? OverlayType.rectangle
                                : OverlayType.circle;
                      });
                    },
                  ),
                  ElevatedButton(
                    child: const Text('Crop image'),
                    onPressed: () async {
                      Directory tempDir = await getTemporaryDirectory();
                      final path = tempDir.path;
                      final fileName = 'crop_${DateTime.now().millisecondsSinceEpoch}';

                      final file = await Cropper.crop(
                        cropperKey: _cropperKey,
                        path: path,
                        fileName: fileName,
                      );
                      setState(() {
                        _croppedFile = file;
                      });
                    },
                  ),
                  IconButton(
                    onPressed: () {
                      setState(() => _rotationTurns--);
                    },
                    icon: const Icon(Icons.rotate_left),
                  ),
                  IconButton(
                    onPressed: () {
                      setState(() => _rotationTurns++);
                    },
                    icon: const Icon(Icons.rotate_right),
                  ),
                ],
              ),
              const SizedBox(height: 16),
              if (_croppedFile != null) Padding(
                padding: const EdgeInsets.all(36.0),
                child: Image.file(_croppedFile!),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
